File: project/backend/app/schemas/queue.py
from datetime import datetime
from pydantic import BaseModel


class QueueJoinResponse(BaseModel):
    queue_position: int
    total: int
    message: str


class QueueLeaveResponse(BaseModel):
    message: str


class QueueStatusResponse(BaseModel):
    in_queue: bool
    queue_position: int | None = None
    total: int | None = None
    is_notified: bool = False
    accept_until: datetime | None = None

File: project/frontend/src/api/queue.ts
const API_BASE = import.meta.env.VITE_API_URL ?? 'http://localhost:8000'

export interface QueueJoinResponse {
  queue_position: number
  total: number
  message: string
}

export interface QueueLeaveResponse {
  message: string
}

async function authFetch<T>(path: string, token: string, method: string): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    method,
    headers: { Authorization: `Bearer ${token}` },
  })
  const data = await res.json()
  if (!res.ok) throw new Error(data.detail ?? '요청 실패')
  return data
}

export function joinQueue(token: string): Promise<QueueJoinResponse> {
  return authFetch('/queue/join', token, 'POST')
}

export function leaveQueue(token: string): Promise<QueueLeaveResponse> {
  return authFetch('/queue/leave', token, 'DELETE')
}